text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,491 @@
+- Feature Name: Arbitrary Self Types 2.0
+- Start Date: 2023-05-04
+- RFC PR: [rust-lang/rfcs#3519](https://github.com/rust-lang/rfcs/pull/3519)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Allow types that implement the new `trait Receiver<Target=Self>` to be the receiver of a method.
+
+# Motivation
+[motivation]: #motivation
+
+Today, methods can only be received by value, by reference, or by one of a few blessed smart pointer types from `core`, `alloc` and `std` (`Arc<Self>`, `Box<Self>`, `Pin<P>` and `Rc<Self>`).
+
+It's been assumed that this will eventually be generalized to support any smart pointer, such as an `CustomPtr<Self>`. Since late 2017, it has been available on nightly under the `arbitrary_self_types` feature for types that implement `Deref<Target=T>` and for raw pointers.
+
+This RFC proposes some changes to the existing nightly feature based on the experience gained, with a view towards stabilizing the feature in the relatively near future.
+
+## Motivation for the arbitrary self types feature overall
+
+One use-case is cross-language interop (JavaScript, Python, C++), where other languages' references can’t guarantee the aliasing and exclusivity semantics required of a Rust reference. For example, the C++ `this` pointer can't be practically or safely represented as a Rust reference because C++ may retain other pointers to the data and it might mutate at any time. Yet, calling C++ methods intrinsically requires a `this` reference. With "arbitrary self types", smart pointer types can be created which obey foreign-language semantics and can be used in safe Rust code:
+
+```rust
+#[repr(transparent)]
+#[derive(Clone)]
+/// A C++ reference. Obeys C++ reference semantics, not Rust reference semantics.
+/// There is no exclusivity; the underlying data may mutate, etc.
+pub struct CppRef<T: ?Sized> {
+    ptr: *const T,
+}
+
+impl<T: ?Sized> Receiver for CppRef<T> {
+    type Target = T;
+}
+
+// generated by bindings generator
+struct ConcreteCppType {
+    // ...
+}
+
+// all generated by bindings generator; mostly calls into C++
+impl ConcreteCppType {
+    fn some_cpp_method(self: CppRef<Self>) {
```

Comment:
So, the "obvious" alternative here is to not separately expose `CppRef` and `ConcreteCppType`, but instead expose `ConcreteCppTypeRef`, and write `self` methods on that. Why do you even want to expose `CppRef` as a separate concept?

Reply:
Two ergonomic reasons:
* `CppRef<T>` itself is a fairly rich type, with its own facilities for adjusting lifetimes, casts, etc. We can put all that into a trait that's implemented by every concrete C++ reference type, but it doesn't feel natural.
* It would separate `ConcreteCppType` from `ConcreteCppTypeRef`. That feels unnatural too. In pure-native Rust, we know that `T` is related to `&T`, or that `T` is related to `Arc<T>`. It doesn't seem reasonable to ask everyone to use entirely new types `RefT` or `ArcT` instead of parameterizing the smart pointer type by the pointee. The same applies in this domain.
@davidhewitt would you like to add anything?

Reply:
> It would separate ConcreteCppType from ConcreteCppTypeRef. 

I assume ConcreteCppType wouldn't show up at all in such a case. Is it common to have APIs that have the same type both by-val and by-ref?

> CppRef<T> itself is a fairly rich type, with its own facilities for adjusting lifetimes, casts, etc. We can put all that into a trait that's implemented by every concrete C++ reference type, but it doesn't feel natural.

Can you elaborate on that? C++ references don't seem to have a ton of structure we can use here.

It might be good to link to some repository where a fully-featured `CppRef` type assuming this RFC is worked out, so that one can get an idea of how these facilities really enable useful safe abstraction. (This wouldn't be the first RFC with such an associated motivating example.) After all, as far as I can tell, the motivation entirely rests on having such a type, so working this out in full detail is crucial to justify why the existing language features do not suffice.

Reply:
> s it common to have APIs that have the same type both by-val and by-ref?

Yes - [C++ objects can even be kept on the Rust stack thanks to some fancy move semantics work](https://github.com/google/moveit) though it's more common to [keep them in abstractions of C++ heap pointer storage types](https://cxx.rs/binding/uniqueptr.html).

> It might be good to link to some repository where a fully-featured `CppRef` type assuming this RFC is worked out

[Here's one for C++](https://github.com/google/autocxx/blob/main/src/reference_wrapper.rs), though this does traitful workarounds for the lack of this RFC. It's already linked from the RFC. I know that this isn't the only C++ interop tool that intends to use this RFC to handle C++ references. @davidhewitt do you have a demo PyO3 equivalent?

Reply:
Thanks for the link. However, I am confused: that `CppRef` has a safe constructor for raw pointers! So clearly it can't be used to build a safe abstraction, as that would then deref raw pointers.

You said earlier that the goal of this RFC is to make it possible to wrap C++ APIs in safe abstractions. That means making C++ methods safe to call. `CppRef` is clearly unsuited for that usecase, anyone implementing a method on `CppRef<Self>` *has to* make that method unsafe.

Reply:
> clearly it can't be used to build a safe abstraction, as that would then deref raw pointers.

It doesn't deref raw pointers _in Rust_. It does pass those pointers back to C++, which might do arbitrary things to them. Imagine that instead of taking a raw pointer, it took a `usize`.

You might argue that _any_ call into C++ is fundamentally unsafe because it might do arbitrary bad things. I agree! I'd argue that the fundamental unsafe-ness of C++ isn't changed by whether we're passing a pointer into it via some Rust code (which doesn't touch or dereference that pointer). So, whether C++ functions are marked as `unsafe` or not is orthogonal to this RFC. It's a [long standing debate](https://steveklabnik.com/writing/the-cxx-debate) and I'd strongly prefer not to relitigate it on this RFC. Even if one believes that all calls into C++ need to be marked `unsafe`, it's still desirable to make as-safe-as-possible wrappers on top of those C++ calls, to try to avoid C++ crashes wherever possible.

Reply:
> It doesn't deref raw pointers in Rust. It does pass those pointers back to C++, which might do arbitrary things to them. Imagine that instead of taking a raw pointer, it took a usize.

Yes, so what?

It is the job of a safe Rust wrapper around a C++ API to know the preconditions of the C++ code, and to check them all on the Rust side. That's what being a safe wrapper *means*. `CppRef` is unsuited for that purpose. And without `CppRef` being suited for that purpose I don't see what the motivation is for this entire RFC.

The cxx debate is different. The cxx debate is about whether such a promise should be made in general by a binding tool. But it is completely accepted that a concrete C/C++/anything function can be wrapped in a safe Rust caller. Even the standard library does it when it calls `read` or `write` or any of the other C functions that it calls.

Currently the RFC claims that it aims to provide a safe wrapper around C++ API, but then fails to actually show that it is useful in achieving that goal.

Reply:
OK. I agree the RFC shouldn't claim it can provide a fully _safe_ wrapper around a C++ API, because (as I think we agree) a tool like bindgen can't make those promises (and this RFC is aimed at such automated tools). I will edit to ensure there are no such claims.

But "safe*r*" is important. In practice, a type like `CppRef<T>` can model C++ reference behavior far better than `&T`.

Overall I believe:

* there's going to be C++/Rust interop
* some of it is going to be achieved by automated binding generators
* C++ types will be represented as Rust types
* people are going to call C++ methods on those types.

Given those assumptions, I believe this RFC (and something like `CppRef`) is a step forward for safety. Right now, in the very widely-used cxx, `&T` is used as references to C++ types which may mutate underneath Rust at any time [0]. `CppRef<T>` is pragmatically less bad.

Let's also not forget that this is not just about C++. The Python community wants this for similar reasons - @davidhewitt might be able to provide an example which doesn't cause quite so much concern as typical C++ unsafety.

[0] = in most cases cxx manages to make the Rust type into a ZST so there's no UB... but not all.

Reply:
It is the "automatically generated" part that I previously was missing, and that lead to us talking past each other for a bit here. Good that we reached the bottom of this!

So, I would suggest you update the motivating example in the RFC to mention all these things -- in particular the aspect about having a tool that translates a C++ reference into *some* Rust type, automatically.

I realize there's still one point I am confused though: I assume such a translation would use Rust `self` whenever there is a method with `this` in C++. But `this` has pointer type, not reference type, right? So when would such auto-generated bindings have `self: CppRef<Self>` anywhere?

> Let's also not forget that this is not just about C++. The Python community wants this for similar reasons - @davidhewitt might be able to provide an example which doesn't cause quite so much concern as typical C++ unsafety.

Sure, happy to hear about other usecases as well. :)

Reply:
Thanks, I'll adjust as you suggest!

Reply:
There's still an open question. :)

"I realize there's still one point I am confused about though: I assume such a translation would use Rust self whenever there is a method with this in C++. But this has pointer type, not reference type, right? So when would such auto-generated bindings have self: CppRef\<Self> anywhere?"

I get that `type& x` in C++ becomes `x: CppRef<type>`, but where does the `self` come in? Is the argument that because `this` is non-null it is basically a reference even though it technically has pointer type?

Reply:
> `this` has a pointer type:

Yes, don't focus too much on the name `CppRef<T>`. In fact, one of the other bindings generation tools wanting this feature plans to call a similar thing `CppPtr<T>`, and it wouldn't surprise me if the ecosystem eventually converges around separate `CppRef<T>` and `CppPtr<T>` types. As you say, `this` is a bit more like a C++ reference in that it's supposed to be non-null (though C++ being C++, sometimes it is null, alas).

The main thing is, we can't use existing Rust types for any of this:
* Rust references have semantics which are incompatible with 
* Without arbitrary self types, we can't call methods on a Rust pointer type. And even if we could, we'd like to wrap them up in a newtype wrapper (to add functionality, check invariants wherever possible, and reduce the risk that users might get these pointers mixed up with pointers to Rust data in unsafe code)

Reply:
@adetaylor @RalfJung sorry to have been slow to reply on this thread. The PyO3 use case can be summarised by a smart pointer `Py<T>` which represents data on the Python side. (For the sake of separating from `CppRef<T>` here, `Py<T>` is `unsafe` to construct, once constructed it exposes an API to call Python types.)

The type `T` is usually opaque and we don't want `Py<T>` to deref to `T` for various reasons. We would still like to be able to use `Py<T>` as a receiver for methods defined for `T`.

For example, we have `Py<PyAny>`, which has methods applicable for any Python object. We also have `Py<PyList>`, which should have the _same_ methods as `Py<PyAny>` _plus_ additional methods which are only meaningful to lists.

In current stable Rust we can model this using traits, however arbitrary self types looks like a much cleaner way to define the complete interfaces for these types `T`. At the moment associated methods like `PyList::new` are the only ones which get defined on `T`, and the rest must be defined in traits. This spilts documentation, for example.

There are various discussion threads that I can link you to if you want to read into this in more detail. The PyO3 API is going under a redesign at the moment to transition towards this `Py<T>` + traits design, so documentation will not align with what's described here.

Reply:
Thanks for updating the RFC. :)

Reply:
If I may add, in PyO3 you can have `Py<T>` where `T` is a user-defined type, which imo is an even stronger case for arbitrary self types

Reply:
Yes, very true. Beyond just wanting this for ourselves, it would be very attractive for users to be able to add methods to their user-defined Python types.